### PR TITLE
Replaced my own DFT algorithm by a third party module FFT (improves greatly perfs)

### PR DESCRIPTION
**Pull request by BenPyton**
@hartwork How to update the github actions to pull in the submodules?
I see the checkout is that:
```
- name: Checkout Git branch
        uses: actions/checkout@v3
```

### CODE REVIEW

**Comment by BenPyton** [comment]
> How to update the github actions to pull in the submodules?

Found out how to do it.

**Comment by BenPyton** [comment]
@hartwork If you have some time, can you tell me if the program (with those changes) runs correctly on linux please?

**Comment by hartwork** [comment]
@BenPyton I just tried and you won't like to hear this: There is a significant drop in output quality. What I did is record a ~5 seconds video before and after while producing mono(!) input from "playing two finger drums right on the microphone". For the videos:

- [`recording-master.mp4`](https://user-images.githubusercontent.com/1577132/218284977-c88fe99d-00b7-429d-b49b-af5c755bc8b4.mp4) (before)
- [`recording-fft-real.mp4`](https://user-images.githubusercontent.com/1577132/218284982-64e7c18c-7a4b-4842-9441-ea14cad066ad.mp4) (after)

What I noticed is that:
- After these changes only the lowest 2 bands show action, but it was many more previously.
- After these changes there is some stray rendering off the diagonal (e.g. near 3 seconds), previously it was only the diagonal as expected with pure mono input.

Additional notes:
- Recording was done with Open Source [SimpleScreenRecorder](https://github.com/MaartenBaert/ssr) and PulseAudio on Linux.
- While I like this microphone, you'd probably consider it low quality, that's okay and expected. It's [this Yanmai USB one](https://smile.amazon.de/gp/product/B07XKXTN95/) to be precise.

---

I would also like to note that if OsciView was (or will become) software to be packaged in Linux and/or homebrew at some point, it would be nice to depend on an FFT library that is widely package/available and fft-real doesn't seem to be one of those, it's literally not packaged anywhere.
Here's some examples of FFT libs are readily available on Gentoo:

- `dev-libs/pocketfft`
  -  Homepage: https://github.com/mreineck/pocketfft/
  -  Description: PocketFFT for C++
- `sci-libs/djbfft`
  -  Homepage: https://cr.yp.to/djbfft.html
  -  Description: Extremely fast library for floating-point convolution
- `sci-libs/fftw`
  -  Homepage: https://www.fftw.org/
  -  Description: Fast C library for the Discrete Fourier Transform
- `sci-libs/hipFFT`
  -  Homepage: https://github.com/ROCmSoftwarePlatform/hipFFT
  -  Description: CU / ROCM agnostic hip FFT implementation
- `sci-libs/kissfft`
  -  Homepage: https://github.com/mborgerding/kissfft
  -  Description: A Fast Fourier Transform (FFT) library that tries to Keep it Simple, Stupid
- `sci-libs/nfft`
  -  Homepage: https://www-user.tu-chemnitz.de/~potts/nfft/
  -  Description: library for nonequispaced discrete Fourier transformations
- `sci-libs/rocFFT`
  -  Homepage: https://github.com/ROCmSoftwarePlatform/rocFFT
  -  Description: Next generation FFT implementation for ROCm

At least one of these explicitly mentions "float". Just an idea.

---

The current use of Git submodules will make some users run into compile errors when they first try to build it, and then find they didn't fetch submodules the hard way. It would be nice to have the build system check for that folder being filled and produce a helpful message OR use one of CMake's builtin features for fetching dependency code from Git repositories, e.g. see https://cmake.org/cmake/help/latest/guide/using-dependencies/index.html#downloading-and-building-from-source-with-fetchcontent as an alternative to submodules. Just an idea.

**Comment by BenPyton** [comment]
@hartwork Thanks for your feedback.

First of all, I would like to say that I made those changes some times ago. I kept it locally because I was too lazy to commit the changes.
Currently, I don't plan to develop or maintain actively the program (hence the `Not Maintained` in the readme), but I encourage you to do so. 🙂

---

> After these changes only the lowest 2 bands show action, but it was many more previously.

Yes because it is now more accurate. I think it depends on 2 things I've improved:
- I use now the [Hann window](https://en.wikipedia.org/wiki/List_of_window_functions#Hann_and_Hamming_windows) I have introduced in the `StereoRecorder`. It is smoothing the start and end of the sample buffer, thus cancelling the high frequencies caused by them.
- The recording of the samples are missing far less audio samples than before. Because I was only getting samples of 8ms windows and when OsciView was at 30FPS (33ms) I was getting only the last 8ms (so getting only 25% of the samples). Now it is storing the samples in a buffer defined by `FFT_SIZE` (1024). So now it will miss samples if the frame time is over `48kHZ/1024 = 46ms (21FPS)`

Here a video I've just made comparing the master branch, the real-fft branch and the foobar2000 software. 
The audio file used in the video [is this one](https://www.youtube.com/watch?v=KFtd-OlNbtw) (The free download is in description)

https://user-images.githubusercontent.com/13402355/218322258-9ee34878-ac35-4615-8c73-c3fc283d9f08.mp4

As you can see, the frequencies from fft-real branch looks a lot more like the ones displayed by foobar2000 (the scale is different though). And it is more stable too than the master branch. Also, there are a lot less flickering in the oscilloscope view.

foobar2000 is using different parameters for the mel scale. I've reversed engineered them and set them in comments in `SpectrumView.cpp:76` and in `FreqView.cpp:62`. However I kept the params from the [mel scale wiki page](https://en.wikipedia.org/wiki/Mel_scale).

> After these changes there is some stray rendering off the diagonal

After you mentioned it I saw them also in my video too. I don't really know why it appeared, but I think a possible lead would be the buffer system I made in the `StereoRecorder`. Maybe there are sometimes some samples offset by one between right and left. It needs more investigation, but I will not do that now.

---

>  it would be nice to depend on an FFT library that is widely package/available and fft-real doesn't seem to be one of those, it's literally not packaged anywhere.

I understand your point of view. However, fft-real is not packaged because it is only code, and is compiled with the software using it. And in fact it **is** available anywhere (any linux distrib, windows or mac) because it is a public github repo. And it is only necessary for those who want to build OsciView. Once build, no package is necessary for fft-real since it is in the program binary.
Here some reasons why I used this one over others I found:
- It uses real samples, not complex. It could be float or double I don't really care (even if double would be better).
- Some classes (the ones I use) allow to set buffer size and sample type at compile time. Thus reducing lots of instructions and memory and getting faster code.
- An FFT is an FFT. the results got in one library should be the same as in another library for the same samples. Maybe the only significant difference is in implementation, where there are some lib faster than others. (This one is fast enough in my opinion)

I know there are plenty fft library, and if you think it is worth the time and effort to change for another one, then I encourage you to do pull request. 🙂 

---

> The current use of Git submodules will make some users run into compile errors when they first try to build it

You're right, I will update the cmake to address that then! 🙂

**Comment by hartwork** [comment]
Hi @BenPyton when I wrote "significant drop in output quality" I knew that it was my perception as a user and not objective truth (unless verified) and that it could well be the other way around. Foobar200 could be just as wrong, but since you seem closer to theory things like Hann and Mel than me, maybe you're just right.

Regarding availability, your argument is essentially "bundling" code which is not a good idea in terms of updates and security. Downstream has to explain that to upstream again and again, and so there are articles like https://wiki.gentoo.org/wiki/Why_not_bundle_dependencies by now if you're interested. This happens to be the Gentoo wiki but Debian would be on the very same page with this. In short: distros want to avoid bundles for good reason.

File: CMakeLists.txt [hunk 1]
```diff
@@ -18,6 +18,10 @@ if(WIN32)
     set(SFML_DIR "${SFML_HOME}/lib/cmake/SFML")
 endif()
 
+if(NOT EXISTS "${CMAKE_SOURCE_DIR}/modules/fft-real/ffft")
+    message(FATAL_ERROR "The submodule fft-real was not found. If you use git, run this command \"git submodule update --init --recursive\". If you do not use git, please download the module from https://github.com/andywang0607/fft-real and place it in the \"modules/\" folder")
```
Comment (hartwork) [inline CMakeLists.txt]: @BenPyton this looks nice, but the forward slash will likely feel alien to users on Windows (so I'd maybe drop it) and the end seems to be missing a full stop, would at least look more natural with a full stop. Just an idea.

**Comment by BenPyton** [comment]
@hartwork 

> Regarding availability, your argument is essentially "bundling" code which is not a good idea in terms of updates and security.

As I said earlier, I understand completely your (this) point of view.

However, I don't think that "security" and "update" apply to this lib. It does not access to anything on the Internet or via file access, no code injection, etc. It does not have more security holes than a HelloWorld code. Also, it has not been updated for more than a decade.
(I know those are subjective arguments, and can be refuted, but this is my state of mind for this little personal project used by 2 people.)

Also, I am not a programer on Unix systems, but on Windows. From the article you've linked, you can read this: 

> Especially in Windows, shipping dependencies can be a favor to users to save end users having to manually install dependencies or additional libraries. Without a package manager there is no real-solution to that on Windows anyway. It is tempting when using bundled code on Windows to also bundle on GNU/Linux too. It feels consistent and fits together nicely in the mind of the software author.

I totally agree, Windows is a real mess for this type of reusable code. It is largely lot easier for developpers and users to bundle all code and dlls with the software.

Also, I don't plan to develop further more this software. I made it just for fun.

Nevertheless, if you want to use a packaged FFT lib on linux instead, I will be happy to review and merge a PR. And porting the cmake to Windows too (to bundle the package :wink:).

File: CMakeLists.txt (hunk 1 above)
Comment (BenPyton) [reply CMakeLists.txt]: @hartwork I presume that users willing to compile this software have some little background on programing. I am a Windows user and I am used to read paths with backslashes as well as forward slashes.

However you're right, I will do some checks to change the path displayed depending on the platform.

Also you're right for the full stop at the end, I'll add it.

**Comment by hartwork** [comment]
@BenPyton I found out that at least part of the artifacts (despite mono input) are from one thread reading data while another thread is writing the same data, as seen in experiments. If all three of `StereoRecorder::onProcessSamples`, `StereoRecorder::updateSamples` and `OsciView::draw` were run in the same thread, there is some chance we would see no artifacts any more, but that is probably not an option. I understand that audio code wants to avoid locks, but to work without locks we'd need some other barriers or more space between the reader and the writer to ensure they don't step on each other's toes.

**Comment by BenPyton** [comment]
Hi @hartwork,

Thank you so much to take time to investigate this issue!

Normally, I have written the code to take into account the different thread, by using 2 buffers for each channel I swap.
Thus, writting and reading should not be done on the same buffers.

However, I have maybe done some wrong logic for that... I need to get back in the code to understand what I did :sweat_smile:

Nevertheless, all this logic is done in `StereoRecorder.cpp`.

I donlt have much time yet, but later (in several days) I will investigate further if you do not do it before.

**Comment by hartwork** [comment]
@BenPyton sounds good, see you! :+1:

**Comment by BenPyton** [comment]
@hartwork I have updated `CMakeLists.txt` to automatically do the git submodule update in the cmake.
If the user don't used git to clone the project, then the message displays saying to download fft-real manually.

Also, I have read my code of `StereoRecorder.cpp`, and you're right.
It is obvious that I did it wrong last time...
The buffer swap should be done in the `updateSamples()` (main thread) and I made some weird things in the `onProcessSamples()` function.

I know how to fix that. If you didn't work on it yet, I will do. The fix should be done in the next days.

> since you seem closer to theory things like Hann and Mel than me, maybe you're just right.

About FFT and hanning window to smooth samples, I found back the link I used when I wrote the code.
If you're interested to know more about smoothing windows for signal processing, here the link:
https://community.sw.siemens.com/s/article/window-types-hanning-flattop-uniform-tukey-and-exponential

**Comment by hartwork** [comment]
@BenPyton I've only been debugging, please feel free to go ahead with your fix. Thanks for the link!

**Comment by BenPyton** [comment]
> After these changes there is some stray rendering off the diagonal, previously it was only the diagonal as expected with pure mono input.

@hartwork  Fixed by the last commit. Tell me if you still have some rendering issues. 🙂
If all is good, then I will merge the PR.

File: src/StereoRecorder.cpp [hunk 2]
```diff
@@ -45,61 +45,49 @@ void StereoRecorder::updateSamples()
...
-		{
-			m_leftDataBuffer[i] = _samples[2 * (nbSample + i)];
-			m_rightDataBuffer[i] = _samples[2 * (nbSample + i) + 1];
-		}
-
-		m_pos = maxSample - nbSample;
-	}
-	else
-		m_pos += nbSample;
+	m_pos = m_bufferOverflow ? 0 : m_pos + nbSample;
```
Comment (hartwork) [inline src/StereoRecorder.cpp]: My vote for adding braces here to save readers wondering about operator precedence.

**Comment by hartwork** [comment]
> Tell me if you still have some rendering issues. :slightly_smiling_face:

@BenPyton I could not see any artifacts when testing this locally with the same finger drum method today — cool! :+1:

File: src/StereoRecorder.cpp (hunk 2 above)
Comment (BenPyton) [reply src/StereoRecorder.cpp]: @hartwork Done!
I've added braces around the condition too, some people prefers that also for readability 🙂

**Comment by hartwork** [comment]
@BenPyton anything left to merge from your point of view?

File: src/StereoRecorder.cpp (hunk 2 above)
Comment (hartwork) [reply src/StereoRecorder.cpp]: Thanks! :+1:

**Comment by BenPyton** [comment]
If there's nothing more on your side, then I merge 😄

**Comment by hartwork** [comment]
@BenPyton I _would_ first need to learn more about FFT, Mel and Hanning to be able to do a _full_ review here. With where I am today, if you're good with status, I'm good with status quo. Let's get this merged :+1:

**Comment by hartwork** [comment]
@BenPyton PS: My own try will be at https://github.com/hartwork/visdriver/blob/e4972e34f0b93f43607a6b0f5b576f46936c9437/src/visualization.c#L54-L57 soon, I hope.

**Comment by BenPyton** [comment]
Yeah, I am not an expert of these domains neither, but it's always better than before those changes 😅
I merge that and like I said, I don't plan to continue the dev on it (until someone has more knowledge about that and wants to improve it with a pull request)

**Pull request merged**